Add `AcquireTokenForApp(scopes)` to `ITokenAcquisition`

jennyf19 opened the pull request:
PR also includes integration tests and lab infrastructure tests 

- first commit has the API changes, @jmprieur please review the XML, as well as the integration test set-up
- second commit has the lab infrastructure added. tests will probably fail unless ran locally.

https://github.com/AzureAD/microsoft-identity-web/issues/26

jennyf19 commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -88,5 +88,16 @@ public interface ITokenAcquisition
             IEnumerable<string> scopes,
             MsalUiRequiredException msalSeviceException);
+
+        /// <summary>
```
@jmprieur no idea what you want said here. please review. :)

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -238,6 +237,32 @@ public class TokenAcquisition : ITokenAcquisition
             return accessToken;
         }
 
+        /// <summary>
```
@jmprieur please review

jennyf19 commented on tests/Microsoft.Identity.Web.Test.Common/TestHelpers/InMemoryTokenCache.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Microsoft.Identity.Web.Test.Common.TestHelpers
+{
+    public class InMemoryTokenCache
```
this is for the lab testing

jennyf19 commented on tests/Microsoft.Identity.Web.Test.LabInfrastructure/Microsoft.Identity.Web.Test.LabInfrastructure.csproj:
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
all the stuff in Microsoft.Identity.Web.Test.LabInfrastructure is copied over from the project in MSAL.NET. only a few small things were changed (mainly visual), but overall, code is the same.

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
+                !string.IsNullOrEmpty(microsoftIdentityOptions.Instance))
```
This needs to be investigated. I found that the applicationOptions did not inherit from the microsoftIdentityOptions in this case, so I had to add values for both in the integration test. Issue tracked here: [https://github.com/AzureAD/microsoft-identity-web/issues/37](https://github.com/AzureAD/microsoft-identity-web/issues/37)

jennyf19 replied:
separate PR and by fixing the test set-up, i no longer have the issue, but we should still look into merging the class. will update the issue.

---
In reply to: [385444897](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385444897) [](ancestors = 385444897)

jennyf19 commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
+
+                    app = ConfidentialClientApplicationBuilder
+                        .CreateWithApplicationOptions(applicationOptions)
+                        //.WithRedirectUri(currentUri)
```
>//.WithRedirectUri(currentUri) [](start = 24, length = 30)

This is another issue. As we calculate the redirectUri here and then use it, if it's not registered, we fail with a null ref. Issue tracked here: [https://github.com/AzureAD/microsoft-identity-web/issues/38](https://github.com/AzureAD/microsoft-identity-web/issues/38)

jennyf19 replied:
working around this for now by creating the httpcontext, but needs to be looked into (hence the issue)

---
In reply to: [385444973](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385444973) [](ancestors = 385444973)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -238,6 +237,32 @@ public class TokenAcquisition : ITokenAcquisition
+
+            // Use MSAL to get the right token to call the API
+            var application = GetOrBuildConfidentialClientApplication();
+            string accessToken;
```
nit: define and assign `accessToken` on the same line.

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -303,6 +328,7 @@ private IConfidentialClientApplication GetOrBuildConfidentialClientApplication()
         {
+            string instance;
             var request = CurrentHttpContext.Request;
             var microsoftIdentityOptions = _microsoftIdentityOptions;
```
Why are these variables needed? Why not just work with the class memebers (_XXX)

jennyf19 replied:
fixed throughout the class

---
In reply to: [385591146](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385591146) [](ancestors = 385591146)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-            IConfidentialClientApplication app = null;
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
```
Please combine the 2 IF statements.

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-            IConfidentialClientApplication app = null;
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
```
I don't have the context around what `applicationOptions` and `microsoftIdentityOptions` are, but it looks to me that it violates the Single Responsability Principle. This class knows too much about options - this is a code smell that you are missing an object whose responsability is to know how to combine these options.

jennyf19 replied:
agree. will do this in a separate PR

---
In reply to: [385592741](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385592741) [](ancestors = 385592741)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-                    .Build();
+                if (!applicationOptions.Instance.EndsWith("/"))
+                    applicationOptions.Instance += "/";
+                instance = applicationOptions.Instance;
```
If both applcationOptions and microsoft Options are defined, your instance will be NULL, probably leading to a null ref.

jennyf19 replied:
n/a

---
In reply to: [385593451](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385593451) [](ancestors = 385593451)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
+                        .WithB2CAuthority(authority)
+                        .Build();
+                }
+                else
```
what about ADFS?

jennyf19 replied:
i made a separate issue for this. 

---
In reply to: [385593891](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385593891) [](ancestors = 385593891)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
+                }
+
+                // Initialize token cache providers
+                _tokenCacheProvider?.InitializeAsync(app.AppTokenCache);
```
Aren't you getting warnings about these lines? As a general rule of thumb, you should always (well 99% of time) `await` async operations. Otherwise you have a race condition here - you return an app and can use it to fetch tokens, but the token cache provider might not have finished initialzing. 

Also, because you don't await, you will ignore possible exceptions in the InitializeAsync.

bgavrilMS replied:
This will of course mean that this private method also becomes Async.

jennyf19 replied:
i wasn't getting warnings but have fixed all of these. good catch.

---
In reply to: [385595862](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385595862) [](ancestors = 385595862)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -446,6 +492,19 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
             }
         }
 
+        private async Task<string> GetAccessTokenFromApplicationCacheAsync(
```
This method name is misleading - in case of a cache miss this goes to Evo. nit: I don't think it's even needed, it's just one line of code.

jennyf19 replied:
have removed. i was hoping to isolate stuff to test, but doesn't make sense at this point.

---
In reply to: [385597157](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385597157) [](ancestors = 385597157)

bgavrilMS commented on tests/Microsoft.Identity.Web.Test.Common/Mocks/MockHttpContextAccessor.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Microsoft.Identity.Web.Test.Common.Mocks
+{
+    public class MockHttpContextAccessor : IHttpContextAccessor
```
Why not use NSubstitute to create mocks on the fly? You will have better control over them.

bgavrilMS commented on tests/Microsoft.Identity.Web.Test.Common/TestHelpers/MsalTestTokenCacheProvider.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class MsalTestTokenCacheProvider : MsalAbstractTokenCacheProvider
+    {
+        public IMemoryCache MemoryCache { get; set; }
```
Please review visibility of all these members. I don't think a setter is needed.

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -238,6 +237,32 @@ public class TokenAcquisition : ITokenAcquisition
+            }
+
+            // Use MSAL to get the right token to call the API
+            var application = GetOrBuildConfidentialClientApplication();
```
There is already a class member named `application`. Please rename this class member to `_application` to avoid confusion.

bgavrilMS commented on tests/Microsoft.Identity.Web.Test.Integration/AcquireTokenForAppIntegrationTests.cs:
```diff
@@ -0,0 +1,138 @@
+        private string _ccaSecret;
+        private readonly ITestOutputHelper _output;
+
+        public AcquireTokenForAppIntegrationTests(ITestOutputHelper output) //test set-up
```
Have a look at NUnit's [SetUp] and [OneTimeSetup] attributes. Some of the setup should be performed once, and some of it for each test: 

AcquireTokenForAppIntegrationTests -> once - I guess ctor is ok, but try to use [OneTimeSetup] as this is the recommended pattern. https://github.com/nunit/docs/wiki/SetUp-and-TearDown
InitializeTokenAcquisitionObjects -> once per test ([SetUp])

jennyf19 replied:
i did look at that, but since we are already using xunit, i just stayed with their recommendations. we can change things over though.

---
In reply to: [385617575](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385617575) [](ancestors = 385617575)

pmaytak replied:
XUnit allows the same behavior. I'll take care of it.

jennyf19 replied:
thanks @pmaytak 

---
In reply to: [385919792](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385919792) [](ancestors = 385919792)

bgavrilMS replied:
Yes, appologies, I mistakenly thought you were using NUnit instead of XUnit.

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-            return app;
+            try
+            {
+                if (microsoftIdentityOptions.IsB2C)
```
This piece of logic is not tested.

jennyf19 replied:
we are missing a lot of tests. they will come in future PRs. 

---
In reply to: [385618752](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385618752) [](ancestors = 385618752)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
-                    .WithRedirectUri(currentUri)
-                    .WithB2CAuthority(authority)
-                    .Build();
+                if (!microsoftIdentityOptions.Instance.EndsWith("/"))
```
This piece of logic is not tested.

jennyf19 replied:
this is removed. not applicable anymore

---
In reply to: [385618845](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385618845) [](ancestors = 385618845)

bgavrilMS commented on tests/Microsoft.Identity.Web.Test.Integration/AcquireTokenForAppIntegrationTests.cs:
```diff
@@ -0,0 +1,138 @@
+            async Task result() =>
+                await _tokenAcquisition.AcquireTokenForAppAsync(TestConstants.s_scopesForUser).ConfigureAwait(false);
+
+            Exception ex = await Assert.ThrowsAsync<MsalServiceException>(result);
```
You can assert the error code as well if you cast directly to the type of exception:

` MsalServiceExceptionex = await Assert.ThrowsAsync<MsalServiceException>(result);`

jennyf19 replied:
yes..thanks.

---
In reply to: [385619799](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385619799) [](ancestors = 385619799)

bgavrilMS commented on tests/Microsoft.Identity.Web.Test/AadIssuerValidatorTests.cs:
```diff
@@ -7,6 +7,7 @@
 using System.IdentityModel.Tokens.Jwt;
 using System.Security.Claims;
 using Xunit;
+using Microsoft.Identity.Web.Test.Common;
```
not needed?

jennyf19 replied:
needed for the testconstants.

---
In reply to: [385620462](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385620462) [](ancestors = 385620462)

jennyf19 commented:
addressed your comments and/or made issues for pending items for future PRs.

---
In reply to: [366302244](https://github.com/AzureAD/microsoft-identity-web/pull/39#issuecomment-366302244) [](ancestors = 366302244)

jmprieur commented on src/Microsoft.Identity.Web/ITokenAcquisition.cs:
```diff
@@ -88,5 +88,16 @@ public interface ITokenAcquisition
             MsalUiRequiredException msalSeviceException);
+
+        /// <summary>
+        /// Acquires a token from the authority configured in the app, for the confidential client itself (in the name of no user)
```
Is "in the name of no user" clear? should we write "not on behalf of a user" ?

jmprieur commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -238,6 +237,34 @@ public class TokenAcquisition : ITokenAcquisition
         }
 
+        /// <summary>
+        /// Acquires a token from the authority configured in the app, for the confidential client itself (in the name of no user)
```
same question as for the interface.

jmprieur commented on tests/Microsoft.Identity.Web.Test.Common/TestHelpers/MsalTestTokenCacheProvider.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class MsalTestTokenCacheProvider : MsalAbstractTokenCacheProvider
+    {
+        public IMemoryCache MemoryCache { get; }
```
Should there be one IMemoryCache per cache key?
Meaning should MemoryCache be a Dictionary<string, IMemoryCache>() ?

jennyf19 replied:
synced off line.

---
In reply to: [385982535](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385982535) [](ancestors = 385982535)

bgavrilMS commented on src/Microsoft.Identity.Web/ClientInfo.cs:
```diff
@@ -29,7 +29,7 @@ public static ClientInfo CreateFromJson(string clientInfo)
-            catch (Exception exc)
+            catch (Exception)
             {
                 throw new ArgumentException($"Failed to parse the returned client info. ", nameof(clientInfo));
```
I know this isn't your code, but ArgumentException is really not the appropriate exception here. `ArgumentException`  tells the developer that they did smth wrong :). Parsing client_info, a piece of JSON returned by EVO, has nothing to do with this. 

Personally I would just let the JSON exception flow out of the deserialize (i.e. remove the try / catch). 

In general, it is a good practice to have your own dedicated exceptions for this project, similar to how MSAL has. I suggest that you review the exception strategy before releasing, as changing exceptions is breaking. CC @jmprieur

jennyf19 replied:
we will discuss. thanks for putting this on our radar.

---
In reply to: [386016125](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r386016125) [](ancestors = 386016125)

bgavrilMS commented on src/Microsoft.Identity.Web/ClientInfo.cs:
```diff
@@ -42,9 +42,9 @@ internal static T DeserializeFromJson<T>(byte[] jsonByteArray)
-            using (var stream = new MemoryStream(jsonByteArray))
-            using (var reader = new StreamReader(stream, Encoding.UTF8))
-                return (T)JsonSerializer.Create().Deserialize(reader, typeof(T));
+            using var stream = new MemoryStream(jsonByteArray);
```
Fancy new C# 8 feature you're `using` there :)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -287,61 +314,67 @@ public async Task RemoveAccountAsync(RedirectContext context)
-                    .WithRedirectUri(currentUri)
-                    .WithAuthority(authority)
-                    .Build();
+                _logger.LogInformation(ex.Message);
```
_logger has a LogError with an exception argument. Seems more appropriate.

jennyf19 replied:
fixed. thanks.

---
In reply to: [386016331](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r386016331) [](ancestors = 386016331)

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -287,61 +314,67 @@ public async Task RemoveAccountAsync(RedirectContext context)
+                }
+
+                // Initialize token cache providers
+                await _tokenCacheProvider.InitializeAsync(app.AppTokenCache).ConfigureAwait(false);
```
+1

bgavrilMS commented on src/Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -287,61 +314,67 @@ public async Task RemoveAccountAsync(RedirectContext context)
         /// <returns></returns>
-        private IConfidentialClientApplication GetOrBuildConfidentialClientApplication()
+        private async Task<IConfidentialClientApplication> GetOrBuildConfidentialClientApplicationAsync()
         {
```
`GetOrCreate` is a pretty standard method name. In any case, this is private method, so it does not matter that much. IMO the more explicit the better.
